### PR TITLE
Added DDS topic class

### PR DESCRIPTION
**Pull request by MiguelCompany**
* Improvements on TopicDescription
* Created Topic class (and its corresponding implementation)
* TopicQos methods moved to TopicImpl

Note: This PR builds on top of #1113

### CODE REVIEW

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 1]
```diff
@@ -363,6 +365,30 @@ const fastdds::dds::SubscriberQos& DomainParticipantImpl::get_default_subscriber
...
 }
 
+ReturnCode_t DomainParticipantImpl::set_default_topic_qos(
+        const TopicQos& qos)
+{
+    if (&qos == &TOPIC_QOS_DEFAULT)
+    {
+        TopicImpl::set_qos(default_topic_qos_, TOPIC_QOS_DEFAULT, true);
+        return ReturnCode_t::RETCODE_OK;
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:375]: Yes, it is, as the default qos is always considered valid, so there is no need to call check_qos for it

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 2]
```diff
@@ -363,6 +365,30 @@ const fastdds::dds::SubscriberQos& DomainParticipantImpl::get_default_subscriber
...
+
+    ReturnCode_t ret_val = TopicImpl::check_qos(qos);
+    if (!ret_val)
+    {
+        return ret_val;
+    }
+
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:382]: This is the `set_default_topic_qos`, which will only apply to the creation of new topics. This means that immutability checks are not required here, IMHO.

File: include/fastdds/dds/topic/Topic.hpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    RTPS_DllAPI const TopicListener* get_listener() const;
+
+    /**
+     * Modifies the TopicListener.
+     * @param listener
+     * @return RETCODE_OK
+     */
+    RTPS_DllAPI ReturnCode_t set_listener(
+            TopicListener* listener);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/dds/topic/Topic.hpp:106]: 👍

File: include/fastdds/dds/topic/qos/TopicQos.hpp [hunk 4]
```diff
@@ -36,85 +39,447 @@ namespace dds {
...
+
+    /**
+     * Setter for TopicDataQosPolicy
+     * @param value
+     */
+    void topic_data(
+            const TopicDataQosPolicy& value)
+    {
+        topic_data_ = value;
+        topic_data_.hasChanged = true;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/dds/topic/qos/TopicQos.hpp:89]: 👍

File: src/cpp/dds/domain/DomainParticipant.cpp [hunk 5]
```diff
@@ -145,17 +145,20 @@ DomainParticipant& DomainParticipant::default_publisher_qos(
...
-//DomainParticipant& DomainParticipant::default_topic_qos(
-//        const dds::topic::qos::TopicQos& qos)
-//{
-//    this->delegate()->set_default_topic_qos(qos);
-//    return *this;
-//}
+DomainParticipant& DomainParticipant::default_topic_qos(
+        const dds::topic::qos::TopicQos& qos)
+{
+    if (this->delegate()->set_default_topic_qos(qos) == ReturnCode_t::RETCODE_INCONSISTENT_POLICY)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/dds/domain/DomainParticipant.cpp:156]: RETCODE_IMMUTABLE_POLICY will never be returned, I have added UNSUPPORTED, though.

File: src/cpp/fastdds/topic/TopicImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        const TopicQos& from,
+        bool first_time)
+{
+    (void)first_time;
+    to = from;
+
+    // Topic Qos is only used to create other Qos, so it can always be updated
+}
+
+
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/topic/TopicImpl.cpp:104]: 👍

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:375]: The same could be applied to `TopicImpl::set_qos`?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1843/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1589/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1837/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1867/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1609/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1861/)

File: include/fastdds/dds/domain/DomainParticipant.hpp [hunk 7]
```diff
@@ -178,6 +182,19 @@ class DomainParticipant : public Entity
             const char* typeName);
 
     // TODO create/delete topic
+    /**
+     * Create a Subscriber in this Participant.
+     * @param qos QoS of the Subscriber.
+     * @param listener Pointer to the listener.
+     * @param mask StatusMask that holds statuses the listener responds to
+     * @return Pointer to the created Subscriber.
+     */
```
Comment (laura-eprosima) [inline include/fastdds/dds/domain/DomainParticipant.hpp:191]: Exchange this comment with the one in line 137

File: src/cpp/dds/topic/Topic.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+//        const dds::domain::DomainParticipant& dp)
+//    : ::dds::core::Reference<detail::Topic>(
+//        new detail::Topic(
+//            dp.delegate().get(),
+//            dp.default_Topic_qos(),
+//            nullptr,
+//            dds::core::status::StatusMask::all()))
+//    , participant_(nullptr)
+//{
+//}
```
Comment (laura-eprosima) [inline src/cpp/dds/topic/Topic.cpp:42]: Uncomment also this function

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 9]
```diff
@@ -547,6 +573,38 @@ Subscriber* DomainParticipantImpl::create_subscriber(
...
+    if (! TopicImpl::check_qos(qos))
+    {
+        logError(PARTICIPANT, "TopicQos inconsistent or not supported");
+        return nullptr;
+    }
+
+    //TODO CONSTRUIR LA IMPLEMENTACION DENTRO DEL OBJETO DEL USUARIO.
+    TopicImpl* topic_impl = new TopicImpl(this, type_support, qos, listener);
+    Topic* topic = new Topic(topic_name, topic_impl, mask);
+
```
Comment (laura-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:600]: Create also a new Instance Handle and GUID for the new Topic
```suggestion
    GUID_t topic_guid = guid();
    do
    {
        topic_guid.entityId = fastrtps::rtps::c_EntityId_Unknown;
        rtps_participant_->get_new_entity_id(topic_guid.entityId);
    } while (exists_entity_id(topic_guid.entityId));
    } while (exists_entity_id(sub_guid.entityId));

InstanceHandle_t topic_handle(topic_guid);
    topic_impl->handle_ = topic_handle;
```

File: src/cpp/fastdds/domain/DomainParticipantImpl.hpp [hunk 10]
```diff
@@ -131,6 +133,22 @@ class DomainParticipantImpl
...
+     * @param type_name Data type of the Topic.
+     * @param qos QoS of the Topic.
+     * @param listen Pointer to the listener.
+     * @param mask StatusMask that holds statuses the listener responds to
+     * @return Pointer to the created Topic.
+     */
+    Topic* create_topic(
+            const std::string& topic_name,
+            const std::string& type_name,
+            const TopicQos& qos,
```
Comment (laura-eprosima) [inline src/cpp/fastdds/domain/DomainParticipantImpl.hpp:148]:
```suggestion
            const TopicQos& qos = TOPIC_QOS_DEFAULT,
```

File: src/cpp/dds/topic/Topic.cpp (hunk 8 above)
Comment (IkerLuengo) [reply src/cpp/dds/topic/Topic.cpp:42]: I had to add `topic_name` and `type_name` parameters for this to work.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1882/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1622/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1875/)

**Pull request merged**